datafusion/core/src/physical_plan/joins/utils.rs

```diff
@@ -693,6 +700,222 @@ impl<T: 'static> OnceFut<T> {
     }
 }
 
+pub(crate) fn need_produce_result_in_final(join_type: JoinType) -> bool {
+    matches!(
+        join_type,
+        JoinType::Left | JoinType::LeftAnti | JoinType::LeftSemi | JoinType::Full
+    )
+}
```

Comment:
I would suggest to add more comments to those methods. Originally they are only used by HashJoin and they are private methods. Now they are used both by HashJoin and NestedLoopJoin.